Implement OpenID Connect-based login

sandhose opened the pull request:
This PR implements OIDC based login through the m.login.sso mechanism.

OpenID Connect is based on OAuth2, so this PR is kinda redundant with #7059. For now, it only supports OpenID Connect-compliant providers with the authorization code flow. This could be expanded to support generic OAuth2 provider, as long as they either support the `openid` scope or the userinfo endpoint to fetch user infos.

This is a work in progress, here are a few things that need to be done:

 - [x] make the authorization, token and jwks endpoints configurable. Right now they are discovered from the discovery document.
 - [x] refresh the jwks periodically. The keys can be rotated by the OIDC Provider
 - ~support implicit and hybrid flows~ – edit: implicit and hybrid flows are for single page apps essentially & give the token in the fragment part of the URL, which can't be retrieved server-side
 - [x] make the user attributes mapping configurable (maybe the same way the SAML provider does)
 - [x] handle errors more nicely
 - [x] write some tests
 - [x] make mypy happy, add typings where it makes sense
 - [ ] maybe handle logouts: the OIDC specs support frontchannel (initiated by the app) and backchannel (initiated by the provider) logouts. For that we would need to save the `id_token` we get from the provider

For the flow to be secure (and prevent CSRF attacks for example), I needed to have a way to carry state through the flow. I chose to use a macaroon token with that state in it.

I needed an URL for the authorization flow callback. As it is not part of the spec, I chose to have that endpoint under `/_synapse/oidc/callback`.

I used `OIDC` as a name pretty much everywhere, but it would make sense to use `oauth2` instead ; haven't really decided.

### Pull Request Checklist

<!-- Please read CONTRIBUTING.md before submitting your pull request -->

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

sandhose commented:
I've polished this PR quite a lot the last few days. CI is passing, and I started writing some tests.

Even though it is not ready for merge, mainly because of error handling, I'd love to have a first review of it.

Here are a few things I'd like to know:

 - are the names I chose alright in general? I used `OIDC` pretty much everywhere, but it could be named `OAuth2`, as it can handle generic OAuth2 providers
 - I used `/_synapse/oidc/callback` as a redirect_uri ; should that be something else?
 - to pass data through the OAuth2 flow, I save a cookie with some session informations encoded in a macaroon token. Is that the right approach?
 - I used `authlib` to handle to handle some of the OAuth2/OIDC parts. I added it in `synapse/python_dependencies.py`. Is there somewhere else I should mention it?
 - I'm not sure how to handle errors properly. Raising `SynapseError` in an async function does not seem to do anything.
 - to map the user infos from to provider to a mxid/display name, I used jinja templates that are configurable through the config file. Is that a good approach, or should I opt for something more similar than what is done in the SAML2 module?

A few things that I'd like to do, but that I might do later in other PRs:

 - maybe refactor a bit with the login via JWT module. This would make the JWT login a lot more flexible, like being able to validate tokens through a public/private keypair instead of a shared secret and map something else than the sub claim for the localpart.
 - sync somehow the display name after the initial registration ; maybe periodically, maybe after each login
 - get the user picture from the provider
 - handle logouts. This implies that we save the access/id/refresh tokens

sandhose commented:
I've added some docs to test some providers and handled errors more nicely. It should be ready for review.

ptman commented:
Excellect. I hope this will get merged soon. Not being familiar with hydra or dex, would testing with keycloak be useful? Keycloak can act as a translator between saml and oidc, which can be useful.

sandhose commented:
> Excellect. I hope this will get merged soon. Not being familiar with hydra or dex, would testing with keycloak be useful? Keycloak can act as a translator between saml and oidc, which can be useful.

Sure. I don't have a keycloak instance available for that, but it should work without problem.
Feel free add instructions on how to try it with Keycloak in `docs/dev/oidc.md`.

Conan-Kudo commented:
@sandhose `OIDC` is perfectly fine. Ipsilon supports OIDC and OAuth2 as separate backends, and calling it OIDC makes it easier to understand what to use it with.

OvermindDL1 commented:
Handling forced logout when the OIDC server mandates a logout would definitely be a wanted feature, too many security issues nowadays to not have that.  ^.^;

richvdh commented:
> Handling forced logout when the OIDC server mandates a logout would definitely be a wanted feature, too many security issues nowadays to not have that. ^.^;

for the avoidance of doubt, this is probably true, but let's keep it as a follow up rather than adding further complexity to this PR.

clokep commented on synapse/rest/client/v1/login.py:
```diff
@@ -96,6 +97,9 @@ def on_GET(self, request):
+            flows.append({"type": LoginRestServlet.TOKEN_TYPE})
         if self.saml2_enabled:
             flows.append({"type": LoginRestServlet.SSO_TYPE})
             flows.append({"type": LoginRestServlet.TOKEN_TYPE})
```
I suspect that these should be if-elif statements as we use below in `register_servlets` (essentially meaning you can only have CAS, SAML, or OIDC enabled). This means we should order them the same as below (CAS, SAML, then ODIC).

ptman replied:
Would it not make sense to be able to support several simultaneously?

clokep replied:
I'm unsure if it would make sense, if it would I think that would be a follow-up. I think it is more important to ensure *something* works if you have a weird config.

clokep commented on docs/sample_config.yaml:
```diff
@@ -1424,6 +1424,99 @@ saml2_config:
+    #
+    # Synapse will look for the following templates in this directory:
+    #
+    # * HTML page to display to users if something goes wrong during the
+    #   authentication process: 'oidc_error.html'.
```
Does this differ dramatically from the `saml_error.html` page? It'd be nice if this used the `sso.template_dir` setting instead of adding another one.

clokep commented on synapse/app/homeserver.py:
```diff
@@ -415,6 +420,10 @@ def start():
 
                 # Check if it needs to be reprovisioned every day.
                 hs.get_clock().looping_call(reprovision_acme, 24 * 60 * 60 * 1000)
+            if hs.config.oidc_enabled:
```
Please add a comment above this about what's being configured (and probably leave an empty line from the section above this).

clokep commented on synapse/handlers/auth.py:
```diff
@@ -1194,12 +1194,32 @@ def generate_delete_pusher_token(self, user_id: str) -> str:
         return macaroon.serialize()
 
-    def _generate_base_macaroon(self, user_id: str) -> pymacaroons.Macaroon:
+    def generate_oidc_session_token(
```
Please add a return type and docstring to this!

clokep commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,452 @@
+        )
+
+    async def handle_oidc_callback(self, request):
+        """Handle an incoming request to /_synapse/oidc/callback
```
Can you expand upon what this does though? The SAML docs aren't great about this, but we recently put some effort into improving the CAS code.

clokep commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,452 @@
+SESSION_COOKIE_NAME = b"oidc_session"
+
+
+class OidcError(Exception):
```
Should this inherit from `SynapseError`?

sandhose replied:
I don't think so: it is only raised in `_exchange_code` to then be caught in `handle_oidc_callback` so we can display the error nicely. Not sure if it's the right approach though

clokep replied:
Ah, I think I missed where this is a caught, I'll need to look again! Thanks!

clokep commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,452 @@
+    pass
+
+
+class OidcHandler:
```
Please add type hints and docstrings to the methods of this class.

clokep commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,452 @@
+
+    @property
+    def _uses_userinfo(self) -> bool:
+        # Maybe that should be user-configurable and not infered?
```
```suggestion
        # Maybe that should be user-configurable and not inferred?
```

clokep commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,452 @@
+        return UserInfo(claims)
+
+    async def handle_redirect_request(self, request, client_redirect_url: bytes) -> str:
+        """Handle an incoming request to /login/sso/redirect
```
Please expand upon what this is doing (in particular it is probably nice to mention that this stores the session in a cookie? I think that wasn't fully standard based on some conversation I saw fly past.)

sandhose replied:
> I think OIDC is fine, putting it into a search engine / Wikipedia brings up pretty quickly how it is. It seems that OIDC is more generic than OAuth2? It might be nice to include (in the newsfragement?) that this includes OAuth2 providers though.

OAuth2 is a bunch of RFCs (mainly [6749](https://tools.ietf.org/html/rfc6749)) that define the authorization flows, but not where the endpoint should be, nor how user infos should work. OIDC defines that. For example, take a look at Google's OIDC discovery document, you'll see all the OAuth2 endpoints they have: https://accounts.google.com/.well-known/openid-configuration

> I'm not sure on this, I'm guessing that unlike SAML / CAS the data can't be pushed through with the client redirect?

Well it can, that's the purpose of the `state` parameter, to persist data throughout the flow. The signed cookie is here to be sure that the request is not intercepted by another malicious app/browser and prevent CSRF attacks. See [Auth0 docs about that](https://auth0.com/docs/protocols/oauth2/oauth-state), they say *“The cookie in which the state value is stored should be signed to prevent attackers to forge it.”*.

> This should work fine, we do this frequently. What was not working about it?

I forget to decorate the handlers with `@wrap_html_request_handler`, that's why it was not working. :)

> I suspect being consistent with the SAML implementation is beneficial. The SAML implementation (of a plug-in) probably provides more flexibility too.

> > * sync somehow the display name after the initial registration ; maybe periodically, maybe after each login
> > * get the user picture from the provider
> 
> I don't think we do this with SAML / CAS currently?

Nope, but that's definitely a thing we'd like to do in my company.

> One piece I see missing in here is handling the UI authentication process that was added in #5667. I think we'll definitely want this, but it might be OK to be added as a follow-up.

I'll look into it.

> Also, do you have any good resources for the flow diagram of OpenID Connect? I think it'd help me understand the various stages as I'm reviewing this! Thanks!

For the flow we are using (the authorization code flow), I think [this](https://auth0.com/docs/flows/concepts/auth-code) is a good resource to understand it. For some of the OIDC specific parts, maybe look at [this page](https://auth0.com/docs/protocols/oidc).

Thanks for this first review, I think I have a lot of comments/docstrings to add in various functions :D 
I'll work on it this afternoon/week-end.

clokep commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,452 @@
+            return
+
+        session = request.getCookie(SESSION_COOKIE_NAME)
+        # Remove the cookie
```
Should the cookie be removed in an error situation?

sandhose replied:
I don't think so, see the comment I left on my last commit (d45a65c).

clokep commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,452 @@
+            cookie,
+            path="/_synapse/oidc",
+            max_age="3600",
+            httpOnly=True,
```
Do we want to set `sameSite` or `secure` here?

ptman replied:
SameSite probably, as older browsers should just ignore it, right? But secure only if served over HTTPS. I think lots of debugging scenarios may have stuff running over plain HTTP

clokep replied:
I think I came to the same conclusions when thinking about it briefly.

sandhose replied:
I had issues with SameSite if set to strict, as the cookie would not be in the callback, as the browser does not send SameSite cookies when redirecting from another host

ptman replied:
Yes, I think it should be Lax, not Strict and absolutely not None

sandhose commented:
> I think OIDC is fine, putting it into a search engine / Wikipedia brings up pretty quickly how it is. It seems that OIDC is more generic than OAuth2? It might be nice to include (in the newsfragement?) that this includes OAuth2 providers though.

OAuth2 is a bunch of RFCs (mainly [6749](https://tools.ietf.org/html/rfc6749)) that define the authorization flows, but not where the endpoint should be, nor how user infos should work. OIDC defines that. For example, take a look at Google's OIDC discovery document, you'll see all the OAuth2 endpoints they have: https://accounts.google.com/.well-known/openid-configuration

> I'm not sure on this, I'm guessing that unlike SAML / CAS the data can't be pushed through with the client redirect?

Well it can, that's the purpose of the `state` parameter, to persist data throughout the flow. The signed cookie is here to be sure that the request is not intercepted by another malicious app/browser and prevent CSRF attacks. See [Auth0 docs about that](https://auth0.com/docs/protocols/oauth2/oauth-state), they say *“The cookie in which the state value is stored should be signed to prevent attackers to forge it.”*.

> This should work fine, we do this frequently. What was not working about it?

I forget to decorate the handlers with `@wrap_html_request_handler`, that's why it was not working. :)

> I suspect being consistent with the SAML implementation is beneficial. The SAML implementation (of a plug-in) probably provides more flexibility too.

> > * sync somehow the display name after the initial registration ; maybe periodically, maybe after each login
> > * get the user picture from the provider
> 
> I don't think we do this with SAML / CAS currently?

Nope, but that's definitely a thing we'd like to do in my company.

> One piece I see missing in here is handling the UI authentication process that was added in #5667. I think we'll definitely want this, but it might be OK to be added as a follow-up.

I'll look into it.

> Also, do you have any good resources for the flow diagram of OpenID Connect? I think it'd help me understand the various stages as I'm reviewing this! Thanks!

For the flow we are using (the authorization code flow), I think [this](https://auth0.com/docs/flows/concepts/auth-code) is a good resource to understand it. For some of the OIDC specific parts, maybe look at [this page](https://auth0.com/docs/protocols/oidc).

Thanks for this first review, I think I have a lot of comments/docstrings to add in various functions :D 
I'll work on it this afternoon/week-end.

clokep commented:
Awesome! A couple more things after re-reading the above:
* Doing display name / user picture sync sounds to me like a follow-up -- this PR is already getting big.
* Let me know if you have questions on the UI authentication process bits, it also could probably be done as a follow-up, but we should at least make sure we don't back ourselves into a corner with the design.

ptman commented:
I think displayname / avatar sync should maybe only be done on user creation or at least make it possible to not sync from the auth source. Often central auth can have a policy of only providing the name on official government ID, while people may want to be called a nickname in chat.

sandhose commented:
@clokep I've resolved some of the remarks you did and added a lot of comments. Those should help you understand how the OIDC flow works. I think it's ready for another review.

auscompgeek commented on synapse/handlers/auth.py:
```diff
@@ -1200,7 +1200,26 @@ def generate_oidc_session_token(
+        complete the SSO login flow.
+
+        Args:
+            state (str): The ``state`` parameter passed to the OIDC provider.
+            nonce (str): The ``nonce`` parameter passed to the OIDC provider.
+            client_redirect_url (str): The URL the client gave when it
+                initiated the flow.
+            duration_in_ms (int): An optional duration for the token in milliseconds.
+                Defaults to an hour.
+
+        Returns:
+            str: A signed macaroon token with the session informations.
```
Listing the types in both the docstring and the function signature is redundant. My understanding is that the type hints are preferred.

clokep replied:
Yes, please list them just as type hints (we're slowly transitioning the code as we touch pieces of it).

auscompgeek commented on synapse/handlers/oidc_handler.py:
```diff
@@ -136,10 +168,29 @@ def _validate_metadata(self):
+            ValueError: if something in the provider is not valid
+
+        Returns:
+            OpenIDProviderMetadata: The providers metadata.
```
```suggestion
            The provider's metadata.
```

auscompgeek commented on synapse/handlers/oidc_handler.py:
```diff
@@ -172,6 +250,31 @@ async def load_jwks(self, force=False) -> Dict[str, List[Dict[str, str]]]:
+                {
+                    'token_type': 'bearer',
+                    'access_token': 'abcdef',
+                    'expires_in': 3599,
```
Doesn't this mean the return type is _not_ `Dict[str, str]`?

clokep replied:
I think the return type should likely be `JsonDict` (found in `synapse.types`). It seems like there's a handful of other spots here that should also be using this instead of `Dict[str, str]`.

auscompgeek replied:
Hm. I would've imagined using a `typing_extensions.TypedDict` would work here, but I guess there's no need to if it's only being used as part of a response.

sandhose replied:
Did not know about typing_extensions ; I wanted to use a TypedDict in the beginning, but that would not have worked with py3.5. That's done.

clokep commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,659 @@
+                return caveat.caveat_id[len(prefix) :]
+        raise Exception("No %s caveat in macaroon" % (key,))
+
+    def _verify_expiry(self, caveat: str):
```
```suggestion
    def _verify_expiry(self, caveat: str) -> bool:
```

clokep commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,659 @@
+        try:
+            macaroon = pymacaroons.Macaroon.deserialize(session)
+        except MacaroonDeserializationException as e:
+            self._render_error(request, "invalid_session", str(e))
```
Is this going to cause an end-user to see a Python exception? There's a few other places that an exception is directly converted to a string, so this goes for those as well.

sandhose replied:
Haven't tried that, but probably. It's sometimes hard to write meaningful human-readable error messages, especially as I'm not a native english speaker. That's why I sometime got lazy for some of those messages that should probably never happen 😛 

I'll try to come up with something for those.

clokep replied:
It might be OK to have a fallback message that gets used in some places ("Something went wrong, retry." sort of thing.)

clokep commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,659 @@
+            method="POST", uri=token_endpoint, headers={}, body=""
+        )
+        headers = {k: [v] for (k, v) in headers.items()}
+        qs = parse_qsl(body, keep_blank_values=True)
+        args.update(qs)
```
This is converting from form-encoded data to arguments? How come this is necessary? Is it just to use `post_urlencoded_get_json`?

sandhose replied:
ClientAuth.prepare from authlib will fill the body/headers depending on what auth method is used (`client_secret_post` fills the body, `client_secret_post` uses the `Authorization` header). Unfortunately, the body is returned as a string, not a dict.

Maybe I should pass the args to `prepare` and use `_http_client.request` instead.

sandhose replied:
And that's what I ended up doing. It makes testing a bit harder, but I think the code exchange part is a bit clearer now.

clokep commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,659 @@
+
+        metadata = await self.load_metadata()
+        jwk_set = metadata.get("jwks")
+        if jwk_set is not None and not force:
```
I'm fairly certain I understand what's happening here, we're checking the metadata for a cached version and returning it unless we're asking to reload it, but it would be nice to have a quick comment above this.

clokep commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,659 @@
+                claims_params=claims_params,
+            )
+
+        claims.validate(leeway=120)  # allows 2 min of clock skew
```
Is the 2 minutes arbitrary or is that a standard value?

sandhose replied:
Kinda arbitrary. I saw this value in the [`authlib` library](https://github.com/lepture/authlib/blob/36e2236b56274879688ecb8019bf1e9bc8e43746/authlib/integrations/base_client/remote_app.py#L200). As per the [OpenID.Core spec, sec 2](https://openid.net/specs/openid-connect-core-1_0.html#IDToken), *Implementers MAY provide for some small leeway, usually no more than a few minutes, to account for clock skew.*

clokep replied:
OK! That seems reasonable then.

clokep commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,659 @@
+        return UserInfo(claims)
+
+    async def handle_redirect_request(
+        self, request, client_redirect_url: bytes
```
```suggestion
        self, request: SynapseRequest, client_redirect_url: bytes
```

(You'll need to import it too.)

clokep commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,659 @@
+            self._provider_metadata.update(metadata_response)
+            self._provider_needs_discovery = False
+
+        self._validate_metadata()
```
Should the validation only be done if `_provider_needs_discovery`? Otherwise it'll happen everytime we call `load_metadata` (which seems like quite a bit).

sandhose replied:
The verification is not very expensive, and `_provider_needs_discovery` is only `True` if we're using provider discovery (`oidc_login.discovery: true` in the HS config).

clokep replied:
Alright. Keeping it simple is probably best anyway.

clokep commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,659 @@
+        if self._provider_needs_discovery:
+            url = get_well_known_url(self._provider_metadata["issuer"], external=True)
+            metadata_response = await self._http_client.get_json(url)
+            # TODO: maybe update the other way around to let user override some values?
```
Does this need additional thought?

sandhose replied:
I think that doing it the other way around was a lot harder when I wrote this comment, because there were a bunch of metadatas that were set not from the config. I'll see if I can do something there.

clokep replied:
If this isn't easy we can just remove the comment. This could always be a future improvement if it is necessary.

clokep commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,659 @@
+        state = generate_token()
+        nonce = generate_token()
+
+        cookie = self._macaroon_generator.generate_oidc_session_token(
```
Shouldn't this be `self._auth_handler`?

sandhose replied:
Should it? I see a bunch of things related to sessions in the auth handler, but those sessions are persisted in-memory and not shared across workers, so a signed macaroon might be a more robust solution, as it is completely stateless server side. 

Or are you saying I should factor out the OIDC session generation/verification in the auth handler?

clokep replied:
Sorry my comment wasn't clear -- what I'm saying is that as far as I can tell this PR adds `generate_oidc_session_token` to the `AuthHandler` class, which gets stored as `self._auth_handler` unless I'm missing something.

clokep replied:
Doh, I misread the file and missed an additional class definition in the center. Please ignore this. 😄

sandhose commented:
I've extracted out the user mapping logic into a module dynamically loaded from the config. One thing I'm not entirely sure is if the `map_user_attributes` method should be async? It might make sense to want to fetch some user attributes somewhere asynchronously, maybe using the `Token` we just got from the provider.

I resolved most of the conversations in this PR, thanks @auscompgeek and @clokep for the reviews!

It might soon be time I rebase this PR since there are few conflicting files with develop. I'll wait the next review for that.

I should be able to do another round on Tuesday, so it would be awesome to have another review before that :)

sandhose commented:
I had to rebase on develop and force-push because tests would fail because of merging conflicts, sorry about that.

The interesting commits start at `2fa9958 | OIDC login: custom modules for user mappings`.

auscompgeek commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,829 @@
+
+env = Environment(finalize=jinja_finalize)
+
+JinjaOidcMappingConfig = TypedDict(
```
Is there a particular reason this is a TypedDict rather than a class?

sandhose replied:
Not really. Would it be better? I'm not really a python developer, so if you think that would make more sense, I can change that.

auscompgeek commented on synapse/http/client.py:
```diff
@@ -434,6 +436,10 @@ def get_json(self, uri, args={}, headers=None):
 
             ValueError: if the response was not JSON
         """
+        actual_headers = {b"Accept": [b"application/json"]}
```
Hmm. I wonder whether this should also be setting the User-Agent.

sandhose replied:
That was a discussion on #synapse-dev:matrix.org. I initially had those in the handler because some providers expected to have that header and I thought it would make sense to have them by default on all `*_json` methods in the http client.

sandhose replied:
Oh, nevermind. Did not read your comment properly. That's actually done by `SimpleHttpClient.request`.

clokep commented on docs/sample_config.yaml:
```diff
@@ -1512,22 +1512,39 @@ oidc_config:
+      # module. This section will be passed as a Python dictionary to the
+      # module's `parse_config` method.
+      #
+      # Bellow is the config of the default mapping provider, based on Jinja2
```
```suggestion
      # Below is the config of the default mapping provider, based on Jinja2
```

clokep commented on docs/sample_config.yaml:
```diff
@@ -1512,22 +1512,39 @@ oidc_config:
+        # Jinja2 template for the localpart of the MXID
         #
-        #localpart: '{{ user.preferred_username }}'
+        localpart: "{ user.preferred_username }"
```
These braces should be doubled, I think?

sandhose replied:
Yup. And it should be `*_template`. Will fix that.

clokep commented on docs/sample_config.yaml:
```diff
@@ -1458,6 +1458,94 @@ saml2_config:
+        #
+        #subject_claim: "sub"
+
+        # Jinja2 template for the localpart of the MXID
+        #
+        localpart: "{ user.preferred_username }"
+
+        # Jinja2 template for the display name to set on first login. Optional.
+        #
+        #display_name: "{ user.given_name } { user.last_name }"
```
It seems these had `_template` added to the end of them in the code, but not in the sample configs or the documentation.

I do have a fear that using jinja here is an added layer of complexity that is unnecessary. Is being able to specify a different module enough here?

sandhose replied:
I think that Jinja templates make sense because 
 - it's already used in the HTML/email templates
 - it makes it easy to express some complex attributes combination ; given + family name is a good example for that
 - we could easily provide a bunch of builtin filters : `hexencode`, `dotreplace` (as seen in the SAML handler), a filter to extract the local part of a mxid/email address, etc.

I think the Jinja mapping provider is flexible enough to cover most of the use cases, but a simpler default provider would cover a lot less cases, meaning admins would have to write a module (with all the maintenance this applies) almost all the time

A solution between the two would be to have two mappers builtin: a simpler one that just extracts claims (as the SAML default one does), probably as default, and a Jinja-based one.

clokep commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,840 @@
+        )
+        headers = {k: [v] for (k, v) in headers.items()}
+
+        # Do the actual request
+        # We're not using the SimpleHttpClient util methods as we don't want to
+        # check the HTTP status code and we do the body encoding ourself.
```
We don't check the status code because the important thing is whether there's an `"error"` key in the response?

Is this true for like a 500 response as well or are we just expecting 200 or 4xx?

sandhose replied:
Reading through the OAuth2 RFC ([RFC6749 sec 5.2](https://tools.ietf.org/html/rfc6749#section-5.2)), it seems like errors should be 400 or 401 codes, while success should always be 200. Grepping through Keycloak's code, it seems that they are also using 403 codes.

I think in all cases, we should check for 5xx errors and nicely show/log those errors. For other cases, either: 

 - check the status code, and if it's a 2xx, assume there is no error and don't check for the `error*` fields. In case of a misbehaving provider, it will throw later in the code, resulting to a `M_UNKNOWN` error (which should never happen).
 - check for the `error` field, regardless of the status code
 - check both, and throw if the provider is misbehaving (e.g. a 200 with a "error" field or a 4xx without one)

clokep replied:
I think we probably want to reject all 500 errors, it is likely that those won't even be a JSON response (although I guess an error would be thrown earlier in the process in that case), so maybe this is OK.

sandhose replied:
That's done. I'm checking that http status code & error field matching. In case of a 5xx, I *try* to parse the JSON and extract the "error" field, but if it fails, it returns a generic server error.

This part is heavily commented and all branches are tested, so you should not have trouble figuring what it is doing

clokep commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,840 @@
+            The provider's metadata.
+        """
+        # If we are using the OpenID Discovery documents, it needs to be loaded once
+        # FIXME: should there be a lock here?
```
Do we think this is necessary?

sandhose replied:
When I wrote this, the metadata was loaded on the fly on the first request and not on startup. In that scenario, two requests could come at the same time and both request would load & cache the provider metadata at the same time. I *think* that now that I'm loading the metadata once on startup it *should* not have a chance to load twice at the same time. I could be wrong, I'm definitely not aware enough of how async python works.

I left this comment in case periodic metadata reloading was implemented.
There could also be the same kind of problems with the JWKS loading. But I don't know if there are python tools to tests race conditions like that.

clokep replied:
We talked a bit about this in #synapse:matrix.org and I *think* we don't need a lock here if we're loading this on start. Regardless, I think the worst case that could happen is we try to fetch the metadata multiple times.

clokep commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,840 @@
+
+        return self._provider_metadata
+
+    async def load_jwks(self, force=False) -> JWKS:
```
```suggestion
    async def load_jwks(self, force: bool = False) -> JWKS:
```

clokep commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,840 @@
+        request.redirect(uri)
+        finish_request(request)
+
+    async def handle_oidc_callback(self, request) -> None:
```
```suggestion
    async def handle_oidc_callback(self, request: SynapseRequest) -> None:
```

clokep commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,840 @@
+            "Retrieved user attributes from user mapping provider: %r", attributes
+        )
+
+        if attributes["localpart"] == "":
```
```suggestion
        if not attributes["localpart"]:
```

clokep commented:
> I've extracted out the user mapping logic into a module dynamically loaded from the config. One thing I'm not entirely sure is if the `map_user_attributes` method should be async? It might make sense to want to fetch some user attributes somewhere asynchronously, maybe using the `Token` we just got from the provider.

There was a little chatter about this in #synapse-dev:matrix.org and the conclusion was that it shouldn't cause an issue to do that and is somewhat annoying to change later on. Also, if there's a bunch more work in order to do that...then we shouldn't worry about it IMO.

I was able to test this with Auth0 and it worked well! 👍 There's a bunch of outstanding comments still (a few are a bit older, by the way). I've resolved the comments that I see as taken care of.

sandhose commented:
@clokep Thanks for the review! I'll work on it tomorrow. I replied to a few of your comments.

maxklenk commented:
Awesome work. To validate how the PR is working I set up OAuth with Hydra and Twitch. 
It took a while to find the `client_auth_method` option, because the returned errors/logging didn't help a lot. 

Please add the Twitch configuration to your examples.

### Twitch

1. Setup a developer account on [Twitch](https://dev.twitch.tv/)
2. Obtain the OAuth 2.0 credentials by [creating an app](https://dev.twitch.tv/console/apps/)
3. Add this OAuth Redirect URL: `[synapse base url]/_synapse/oidc/callback`

```yaml
oidc_config:
   enabled: true
   issuer: "https://id.twitch.tv/oauth2/"
   discover: true
   client_id: "your-client-id" # TO BE FILLED
   client_secret: "your-client-secret" # TO BE FILLED
   client_auth_method: "client_secret_post"
   scopes:
     - openid
   user_mapping_provider:
     config:
       localpart: '{{ user.preferred_username }}'
       display_name: '{{ user.name }}'
```

sandhose commented:
@maxklenk Thanks for testing! I added a test to check if the provider supports the client auth configured for token exchange. It should crash with a meaningful error message now if the config is invalid.

@clokep I think we're getting close of this PR being done! I tried to keep the last few commits atomic and explicit in what they do, so you might want to check that out to see what I've done since last review.

I checked that errors where the exception messages were displayed are actually human-readable, and it seems to be always the case.

I've tested more code paths in functions that were already tested. A few places that are *not* covered by unit tests:

 - the whole `JinjaOidcMappingProvider` (but there is no fancy logic in there)
 - `OidcHandler._parse_id_token`,  because mocking JWT signing is not fun
 - `OidcHandler._fetch_userinfo`, because mocking requests is not fun either (although a lot more feasible)
 - `OidcHandler._map_userinfo_to_user`, because it has a lot of interactions with the datastore, and I don't know how to test those

I've added a few logging statements. Check those and tell me if the error levels make sense.

I did nothing about the Jinja provider. If you want me to add a simpler default provider, it's quite quick to do.

There is no user documentation for now on this feature. Sample config is I think explicit enough for people familiar with OIDC & dev docs has enough sample configs in it :p

ptman commented:
Anything relevant to the security in the RFC or comments at https://news.ycombinator.com/item?id=23080240 ?

sandhose commented:
> Anything relevant to the security in the RFC or comments at https://news.ycombinator.com/item?id=23080240 ?

Most of best practices are implemented here. 
We're using the code authorization flow with nonce & state params, both verified and stored signed in cookies. This mitigates 4.1, 4.2, 4.3, 4.5, 4.6 and 4.7. 
In case of an OIDC-compliant provider, we're extensively checking the content of the ID token. This mitigates 4.4.
Because we don't use access tokens nor refresh tokens to access resources, we're not subject to 4.8, 4.12 and 4.13.
The callback is not an open redirector and we're passing the code through query params, so we're not subject to 4.9
4.10, 4.11 and 4.14 are for authorizations servers, not relying parties, so nothing that can be done on our side.

Things that could be implemented to further improve security:

- [PKCE](https://auth0.com/docs/flows/concepts/auth-code-pkce). That would further consolidate the code exchange flow.
- supporting the `private_key_jwt` and `client_secret_jwt` client auth methods, to avoid leaking client secrets if somehow the network between synapse and the provider is compromised (more informations [here](https://medium.com/@darutk/oauth-2-0-client-authentication-4b5f929305d4)
- Storing access token and check for their revocation periodically (and support back/front-channel logouts) to allow ending synapse sessions from the provider

clokep commented:
@sandhose There's a packaging error (unrelated to this PR), I think you'll need to merge in develop or rebase to fix this.

babolivier commented:
> @sandhose There's a packaging error (unrelated to this PR), I think you'll need to merge in develop or rebase to fix this.

Merging develop in should be fine fwiw

sandhose commented on synapse/handlers/oidc_handler.py:
```diff
@@ -0,0 +1,934 @@
+        uri = prepare_grant_uri(
+            authorization_endpoint,
+            client_id=self._client_auth.client_id,
+            response_type="code",
```
The response type itself is used in two places: here and in the metadata validation. I don't think it makes sense to abstract that out, as we will probably never support other flows

sandhose commented on synapse/config/oidc_config.py:
```diff
@@ -0,0 +1,177 @@
+              # module. This section will be passed as a Python dictionary to the
+              # module's `parse_config` method.
+              #
+              # Bellow is the config of the default mapping provider, based on Jinja2
```
It seems so. grepping through the codebase, it seems like I'm not the only one [making that typo](https://github.com/matrix-org/synapse/blob/207b1737ee0acd226359d59ce3b7f7d46111b1c8/synapse/res/templates/notice_expiry.html#L33) :p

clokep replied:
🤦 #7449 good find!

sandhose commented on tests/handlers/test_oidc.py:
```diff
@@ -0,0 +1,576 @@
+
+
+@contextmanager
+def metadata_edit(handler, values):
```
Did not know about patch.dict, thanks for the suggestion!

sandhose commented:
Phew, that should be it.

I rebased on develop

 - for the packaging issue in CI
 - change the commit author with my work email
 - edit the changelog entry (please tell me if it's OK)

So the only commit I really changed is `OIDC login: add changelog entry`, 7dccd63dc

---

Anyway, the new commits start at `OIDC login: make the user attribute mapping async`, 9bd40d1fd.

What was done (in commit order):
 - making the attribute mapper async, and add the token as parameter. I think it could make sense in some providers that might want to do authenticated requests to fetch some user attributes (9bd40d1fd)
 - changed the metadata validation a bit. Went from `if "foo" in metadata` to `metadata.get("foo") is not None`, which also simplified the tests a bit. (eace065c7)
 - refactored macaroon generation and verification ; so the MacaroonGenerator in auth_handler is now untouched by this PR (b3e7b6c22)
 - added docstrings to tests. Not sure about the format, because I had trouble to find tests with docstrings (cfa177cbf)
 - some misc fixes based on the last review (0f5b4fdad)

ptman commented:
Excellent! Will this make it to 1.13 or will it wait for 1.14?

clokep commented:
@ptman This will wait for 1.14, we'll be cutting an RC for 1.13 soon™️.

clokep commented:
Thanks again for all your work on this! I believe we had mentioned a couple of follow-ups above, hopefully we have a good basis now to implement those!

roderikelgodo commented:
Hi. Thank you very much for such a great and hard work. I'm waiting the final release of this "OpenId connect-base login" implementation. Meanwhile I'm testing the branch develop with Dex, using Riot as a native ubuntu client. Everything seems to be working fine: authentication using an internal ldap or github works great (creating users in the database, etc). But the very last part of the process gets stuck in "Go to your browser to complete Sign In" (Riot part), and the "I trust this address" (web part) link that opens the Riot app with the "Go to your browser ...". So it keeps in a loop.
![image](https://user-images.githubusercontent.com/19209191/83055244-e270e280-a053-11ea-923c-bc874f57f107.png)
If I can be of any help for testing, count on that!

clokep commented:
@roderikelgodo Are you using Riot Desktop? I think you're hitting vector-im/riot-web#13719. If it isn't that, please open a new issue or join #synapse:matrix.org.

roderikelgodo commented:
@clokep thank you very much for the answer. Yes, it is [vector-im/riot-web#13719](https://github.com/vector-im/riot-web/issues/13719). After doing other test with other users, passing through Riot Web (and not through Riot Desktop), everything works fine.

mlwalsh1228 commented:
Hello! This feature works so well! I'm using the standard Auth0 authorization code flow. There are a few things I had to modify to get it working, so please correct me if I've not configured something correctly!

1. `_synapse/oidc` was not able to be reached, and I've scanned much documentation to see if I needed to enable something to get the `_synapse` rest api working. I've modified the few places in code to be `_matrix/oidc` and this feature works perfect, but I must be doing something wrong?
2. Setting up the encryption keys through Riot will require you to enter your password as one of the steps, which there isn't one in this case. I've checked upcoming changes to Riot, but did not see the handling for that?

So sorry to waste your time if I've done something incorrect or missed an upcoming change! Let me know if I should file an issue for either of these.

clokep commented:
@mlwalsh1228 Do you have a reverse proxy that is blocking access to things not at _matrix?

Anyway, I'm going to lock this issue, please don't use closed PRs for questions. If you believe there's a bug please open a separate issue, for support requests please use #synapse:matrix.org.
